File: src/src_platform_uwp/fmg/uwp/utils/Cast.cs
using fmg.common;
using fmg.common.geom;

namespace fmg.uwp.utils {

   /// <summary>
   /// Приведение типов от платформонезависемых чистых C# классов fmg.common.* к библиотечным Windows.Foundation классам
   /// </summary>
   public static class Cast {

      public static Size ToFmSize(this Windows.Foundation.Size self) { return new Size((int)self.Width, (int)self.Height); }
      public static Windows.Foundation.Size ToWinSize(this Size self) { return new Windows.Foundation.Size { Width = self.Width, Height = self.Height }; }

      public static SizeDouble ToFmSizeDouble(this Windows.Foundation.Size self) { return new SizeDouble(self.Width, self.Height); }
      public static Windows.Foundation.Size ToWinSize(this SizeDouble self) { return new Windows.Foundation.Size { Width = self.Width, Height = self.Height }; }

      public static Rect ToFmRect(this Windows.Foundation.Rect self) { return new Rect((int)self.X, (int)self.Y, (int)self.Width, (int)self.Height); }
      public static Windows.Foundation.Rect ToWinRect(this Rect self) { return new Windows.Foundation.Rect(self.X, self.Y, self.Width, self.Height); }

      public static RectDouble ToFmRectDouble(this Windows.Foundation.Rect self) { return new RectDouble(self.X, self.Y, self.Width, self.Height); }
      public static Windows.Foundation.Rect ToWinRect(this RectDouble self) { return new Windows.Foundation.Rect(self.X, self.Y, self.Width, self.Height); }

      public static Point ToFmPoint(this Windows.Foundation.Point self) { return new Point((int)self.X, (int)self.Y); }
      public static Windows.Foundation.Point ToWinPoint(this Point self) { return new Windows.Foundation.Point { X = self.X, Y = self.Y }; }

      public static PointDouble ToFmRectDouble(this Windows.Foundation.Point self) { return new PointDouble(self.X, self.Y); }
      public static Windows.Foundation.Point ToWinPoint(this PointDouble self) { return new Windows.Foundation.Point { X = self.X, Y = self.Y }; }

      public static Color ToFmColor(this Windows.UI.Color self) { return new Color(self.A, self.R, self.G, self.B); }
      public static Windows.UI.Color ToWinColor(this Color self) { return new Windows.UI.Color { A = self.A, B = self.B, G = self.G, R = self.R }; }

   }
}

File: FastMines_swing/src/main/java/fmg/swing/utils/Cast.java
package fmg.swing.utils;

import java.awt.GraphicsConfiguration;
import java.awt.GraphicsDevice;
import java.awt.GraphicsEnvironment;
import java.awt.geom.AffineTransform;
import java.util.List;

import fmg.common.Logger;

/** Приведение типов от платформо-независимых чистых Java классов fmg.common.geom.* к библиотечным SWING/AWT классам java.awt.*\java.swing.* */
public final class Cast {

    public static        java.awt.Point       toPoint      (fmg.common.geom.Point p) { return new        java.awt.Point      (p.x, p.y); }
    public static fmg.common.geom.Point       toPoint      (       java.awt.Point p) { return new fmg.common.geom.Point      (p.x, p.y); }
    public static fmg.common.geom.PointDouble toPointDouble(       java.awt.Point p) { return new fmg.common.geom.PointDouble(p.x, p.y); }

    public static java.awt.geom.Point2D.Double toPoint( fmg.common.geom.PointDouble p) { return new java.awt.geom.Point2D.Double(p.x, p.y); }
    public static  fmg.common.geom.PointDouble toPoint(java.awt.geom.Point2D.Double p) { return new  fmg.common.geom.PointDouble(p.x, p.y); }

    public static java.awt.Rectangle         toRect      (fmg.common.geom.Rect       rc) { return new java.awt.Rectangle        (     rc.x,      rc.y,      rc.width,      rc.height); }
    public static java.awt.Rectangle         toRect      (fmg.common.geom.RectDouble rc) { return new java.awt.Rectangle        ((int)rc.x, (int)rc.y, (int)rc.width, (int)rc.height); }
    public static fmg.common.geom.Rect       toRect      (java.awt.Rectangle         rc) { return new fmg.common.geom.Rect      (     rc.x,      rc.y,      rc.width,      rc.height); }
    public static fmg.common.geom.RectDouble toRectDouble(java.awt.Rectangle         rc) { return new fmg.common.geom.RectDouble(     rc.x,      rc.y,      rc.width,      rc.height); }

    public static java.awt.Dimension         toSize      (fmg.common.geom.SizeDouble size) { return new java.awt.Dimension        ((int)size.width, (int)size.height); }
    public static java.awt.Dimension         toSize      (fmg.common.geom.Size       size) { return new java.awt.Dimension        (     size.width,      size.height); }
    public static fmg.common.geom.Size       toSize      (java.awt.Dimension         size) { return new fmg.common.geom.Size      (     size.width,      size.height); }
    public static fmg.common.geom.SizeDouble toSizeDouble(java.awt.Dimension         size) { return new fmg.common.geom.SizeDouble(     size.width,      size.height); }

    public static java.awt.Polygon toPolygon(fmg.common.geom.Region region) {
        java.awt.Polygon polygon = new java.awt.Polygon();
        for (int i=0; i<region.getCountPoints(); i++) {
            fmg.common.geom.Point p = region.getPoint(i);
            polygon.addPoint(p.x, p.y);
        }
        return polygon;
    }

    public static java.awt.Polygon toPolygon(fmg.common.geom.RegionDouble region) {
        java.awt.Polygon polygon = new java.awt.Polygon();
        for (int i = 0; i < region.getCountPoints(); i++) {
            fmg.common.geom.PointDouble p = region.getPoint(i);
            polygon.addPoint((int) p.x, (int) p.y);
        }
        return polygon;
    }

    public static java.awt.Polygon toPolygon(List<fmg.common.geom.PointDouble> region) {
        java.awt.Polygon p = new java.awt.Polygon();
        region.forEach(pd -> p.addPoint((int) pd.x, (int) pd.y));
        return p;
    }

    public static   java.awt.Color toColor(fmg.common.Color clr) { return new java.awt.Color( clr.getR(), clr.getG(), clr.getB(), clr.getA()); }
    public static fmg.common.Color toColor(java.awt.Color   clr) { return new fmg.common.Color(clr.getAlpha(), clr.getRed(), clr.getGreen(), clr.getBlue()); }

    private static final double DISPLAY_DENSITY;
    static {
        double scale = 1; // 100%
        try {
            GraphicsEnvironment ge = GraphicsEnvironment.getLocalGraphicsEnvironment();
            GraphicsDevice gd = ge.getDefaultScreenDevice();
            GraphicsConfiguration gc = gd.getDefaultConfiguration();
            AffineTransform at = gc.getDefaultTransform();
            scale = at.getScaleX();
        } catch (Throwable ex) {
            Logger.error("Can`t find display scale", ex);
        }

        double logicalDpi = scale * 96;
        double density = logicalDpi / 72;
        Logger.info("Curent display density: " + density);

        DISPLAY_DENSITY = density;
    }

    /** Pixels to DPI */
    public static double pxToDp(double px) {
        return px / DISPLAY_DENSITY;
    }

    /** DPI to pixels */
    public static double dpToPx(double dp) {
        return dp * DISPLAY_DENSITY;
    }

}
